File: backend-node/routes/quality.js
const express = require('express');
const router = express.Router();
const { Op } = require('sequelize');
const { QCInspection, DefectLog, DefectType, BatchApproval, sequelize } = require('../models');
const { authenticateToken } = require('../middleware/auth');

// 1. List all QC inspections
router.get('/inspections', authenticateToken, async (req, res) => {
  try {
    const limit = parseInt(req.query.limit) || 100;
    const offset = parseInt(req.query.skip) || 0;
    const { result, wo_id } = req.query;

    const whereClause = {};
    if (result) {
      whereClause.result = result;
    }
    if (wo_id) {
      whereClause.wo_id = parseInt(wo_id);
    }

    const inspections = await QCInspection.findAll({
      where: whereClause,
      order: [['inspection_date', 'DESC']],
      limit,
      offset
    });

    return res.json(inspections);
  } catch (error) {
    console.error('List Inspections Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 2. Create new QC inspection
router.post('/inspections', authenticateToken, async (req, res) => {
  const t = await sequelize.transaction();
  try {
    const { wo_id, po_id, batch_number, quantity_inspected, defects } = req.body;

    // Generate inspection number (QC-YYYYMMDD-XXXX)
    const todayStr = new Date().toISOString().slice(0, 10).replace(/-/g, '');
    const count = await QCInspection.count({ transaction: t });
    const inspection_number = `QC-${todayStr}-${(count + 1).toString().padStart(4, '0')}`;

    const new_inspection = await QCInspection.create({
      inspection_number,
      wo_id,
      po_id,
      batch_number,
      quantity_inspected,
      inspector_id: req.user.user_id,
      result: 'pending'
    }, { transaction: t });

    let total_defects = 0;
    let critical_defects = 0;

    if (defects && defects.length > 0) {
      for (const d of defects) {
        await DefectLog.create({
          inspection_id: new_inspection.inspection_id,
          defect_type_id: d.defect_type_id,
          quantity: d.quantity,
          severity: d.severity,
          location: d.location,
          description: d.description
        }, { transaction: t });

        total_defects += parseFloat(d.quantity);
        if (d.severity === 'critical') {
          critical_defects += parseFloat(d.quantity);
        }
      }
    }

    const defect_rate = quantity_inspected > 0 ? (total_defects / quantity_inspected) * 100 : 0;

    if (critical_defects > 0 || defect_rate > 5) {
      new_inspection.result = 'failed';
      new_inspection.quantity_rejected = total_defects;
      new_inspection.quantity_accepted = quantity_inspected - total_defects;
    } else if (defect_rate > 2) {
      new_inspection.result = 'conditional';
      new_inspection.quantity_accepted = quantity_inspected - critical_defects;
      new_inspection.quantity_rejected = critical_defects;
    } else {
      new_inspection.result = 'passed';
      new_inspection.quantity_accepted = quantity_inspected;
      new_inspection.quantity_rejected = 0;
    }

    await new_inspection.save({ transaction: t });

    await t.commit();

    return res.status(201).json(new_inspection);

  } catch (error) {
    await t.rollback();
    console.error('Create Inspection Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 3. Get inspection details
router.get('/inspections/:inspection_id', authenticateToken, async (req, res) => {
  try {
    const inspection_id = parseInt(req.params.inspection_id);
    const inspection = await QCInspection.findByPk(inspection_id);
    if (!inspection) {
      return res.status(404).json({ detail: 'Inspection not found' });
    }

    const defects = await DefectLog.findAll({
      where: { inspection_id },
      include: [{ model: DefectType, as: 'defect_type' }]
    });

    const insQty = parseFloat(inspection.quantity_inspected || 0);
    const rejQty = parseFloat(inspection.quantity_rejected || 0);

    return res.json({
      inspection,
      defects,
      defect_count: defects.length,
      defect_rate: insQty > 0 ? (rejQty / insQty) * 100 : 0
    });
  } catch (error) {
    console.error('Get Inspection Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 4. Approve inspection
router.put('/inspections/:inspection_id/approve', authenticateToken, async (req, res) => {
  try {
    const inspection_id = parseInt(req.params.inspection_id);
    const { notes } = req.body;

    const inspection = await QCInspection.findByPk(inspection_id);
    if (!inspection) {
      return res.status(404).json({ detail: 'Inspection not found' });
    }

    inspection.approved_by = req.user.user_id;
    inspection.approved_at = new Date();
    if (notes) {
      inspection.remarks = (inspection.remarks || '') + `\nApproval notes: ${notes}`;
    }
    await inspection.save();

    return res.json({
      message: 'Inspection approved successfully',
      inspection_number: inspection.inspection_number,
      result: inspection.result
    });
  } catch (error) {
    console.error('Approve Inspection Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 5. Log defect
router.post('/defects', authenticateToken, async (req, res) => {
  try {
    const { inspection_id, defect_type_id, quantity, severity, location, description } = req.body;

    const defect = await DefectLog.create({
      inspection_id,
      defect_type_id,
      quantity,
      severity,
      location,
      description
    });

    return res.json({
      message: 'Defect logged successfully',
      defect_id: defect.defect_id
    });
  } catch (error) {
    console.error('Log Defect Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 6. List defects
router.get('/defects', authenticateToken, async (req, res) => {
  try {
    const limit = parseInt(req.query.limit) || 100;
    const offset = parseInt(req.query.skip) || 0;
    const { inspection_id, defect_type_id, severity } = req.query;

    const whereClause = {};
    if (inspection_id) {
      whereClause.inspection_id = parseInt(inspection_id);
    }
    if (defect_type_id) {
      whereClause.defect_type_id = parseInt(defect_type_id);
    }
    if (severity) {
      whereClause.severity = severity;
    }

    const count = await DefectLog.count({ where: whereClause });
    const defects = await DefectLog.findAll({
      where: whereClause,
      order: [['logged_at', 'DESC']],
      limit,
      offset
    });

    return res.json({
      total: count,
      defects
    });
  } catch (error) {
    console.error('List Defects Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 7. Defect Types
router.get('/defects/types', authenticateToken, async (req, res) => {
  try {
    const defect_types = await DefectType.findAll({ where: { is_active: true } });
    return res.json({ total: defect_types.length, defect_types });
  } catch (error) {
    console.error('List Defect Types Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 8. Get batch approval status
router.get('/batch-approval/:batch_number', authenticateToken, async (req, res) => {
  try {
    const { batch_number } = req.params;
    const approval = await BatchApproval.findOne({ where: { batch_number } });

    if (!approval) {
      const inspection = await QCInspection.findOne({ where: { batch_number } });
      if (!inspection) {
        return res.status(404).json({ detail: 'Batch not found' });
      }

      return res.json({
        batch_number,
        status: 'not_approved',
        inspection
      });
    }

    return res.json({
      batch_number,
      status: approval.status,
      approval
    });
  } catch (error) {
    console.error('Get Batch Approval Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 9. Approve or reject batch
router.post('/batch-approval/:batch_number', authenticateToken, async (req, res) => {
  try {
    const { batch_number } = req.params;
    const { action, notes, rejection_reason, rework_instructions } = req.body;

    const inspection = await QCInspection.findOne({ where: { batch_number } });
    if (!inspection) {
      return res.status(404).json({ detail: 'No inspection found for this batch' });
    }

    let approval = await BatchApproval.findOne({ where: { batch_number } });

    if (approval) {
      approval.status = action;
      approval.approved_by = req.user.user_id;
      approval.approved_at = new Date();
      approval.notes = notes;
      approval.rejection_reason = rejection_reason;
      approval.rework_instructions = rework_instructions;
      await approval.save();
    } else {
      approval = await BatchApproval.create({
        batch_number,
        inspection_id: inspection.inspection_id,
        status: action,
        approved_by: req.user.user_id,
        approved_at: new Date(),
        notes,
        rejection_reason,
        rework_instructions
      });
    }

    return res.json({
      message: `Batch ${action} successfully`,
      batch_number,
      status: action
    });

  } catch (error) {
    console.error('Approve Batch Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

module.exports = router;

File: backend/app/api/routes/quality.py
"""
Quality API Routes
QC Inspections, Defects, Batch Approvals
"""

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from sqlalchemy import func
from typing import List, Optional
from datetime import datetime
from pydantic import BaseModel

from app.core.database import get_db
from app.core.security import get_current_user
from app.models.auth import User
from app.models.quality import QCInspection, DefectLog, DefectType, BatchApproval, InspectionResult, DefectSeverity

router = APIRouter()

# Schemas
class DefectLogCreate(BaseModel):
    defect_type_id: int
    quantity: float
    severity: str
    location: Optional[str] = None
    description: Optional[str] = None

class InspectionCreate(BaseModel):
    wo_id: Optional[int] = None
    po_id: Optional[int] = None
    batch_number: str
    quantity_inspected: float
    defects: Optional[List[DefectLogCreate]] = []

class InspectionResponse(BaseModel):
    inspection_id: int
    inspection_number: str
    batch_number: str
    result: str
    quantity_inspected: float
    
    class Config:
        from_attributes = True

# Inspection Routes
@router.get("/inspections", response_model=List[InspectionResponse])
async def list_inspections(
    skip: int = 0,
    limit: int = 100,
    result: Optional[str] = None,
    wo_id: Optional[int] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """List all QC inspections"""
    query = db.query(QCInspection)
    
    if result:
        query = query.filter(QCInspection.result == result)
    if wo_id:
        query = query.filter(QCInspection.wo_id == wo_id)
    
    inspections = query.order_by(QCInspection.inspection_date.desc()).offset(skip).limit(limit).all()
    return inspections

@router.post("/inspections", response_model=InspectionResponse, status_code=201)
async def create_inspection(
    inspection_data: InspectionCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Create new QC inspection"""
    # Generate inspection number
    inspection_number = f"QC-{datetime.now().strftime('%Y%m%d')}-{db.query(func.count(QCInspection.inspection_id)).scalar() + 1:04d}"
    
    # Create inspection
    new_inspection = QCInspection(
        inspection_number=inspection_number,
        wo_id=inspection_data.wo_id,
        po_id=inspection_data.po_id,
        batch_number=inspection_data.batch_number,
        quantity_inspected=inspection_data.quantity_inspected,
        inspector_id=current_user.user_id,
        result=InspectionResult.PENDING
    )
    
    db.add(new_inspection)
    db.flush()
    
    # Add defects if any
    total_defects = 0
    critical_defects = 0
    
    for defect_data in inspection_data.defects:
        defect = DefectLog(
            inspection_id=new_inspection.inspection_id,
            defect_type_id=defect_data.defect_type_id,
            quantity=defect_data.quantity,
            severity=defect_data.severity,
            location=defect_data.location,
            description=defect_data.description
        )
        db.add(defect)
        
        total_defects += defect_data.quantity
        if defect_data.severity == DefectSeverity.CRITICAL:
            critical_defects += defect_data.quantity
    
    # Determine inspection result
    defect_rate = (total_defects / inspection_data.quantity_inspected) * 100 if inspection_data.quantity_inspected > 0 else 0
    
    if critical_defects > 0:
        new_inspection.result = InspectionResult.FAILED
        new_inspection.quantity_rejected = total_defects
        new_inspection.quantity_accepted = inspection_data.quantity_inspected - total_defects
    elif defect_rate > 5:  # 5% threshold
        new_inspection.result = InspectionResult.FAILED
        new_inspection.quantity_rejected = total_defects
        new_inspection.quantity_accepted = inspection_data.quantity_inspected - total_defects
    elif defect_rate > 2:  # 2-5% conditional
        new_inspection.result = InspectionResult.CONDITIONAL
        new_inspection.quantity_accepted = inspection_data.quantity_inspected - critical_defects
        new_inspection.quantity_rejected = critical_defects
    else:
        new_inspection.result = InspectionResult.PASSED
        new_inspection.quantity_accepted = inspection_data.quantity_inspected
        new_inspection.quantity_rejected = 0
    
    db.commit()
    db.refresh(new_inspection)
    
    return new_inspection

@router.get("/inspections/{inspection_id}")
async def get_inspection(
    inspection_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get inspection details"""
    inspection = db.query(QCInspection).filter(QCInspection.inspection_id == inspection_id).first()
    if not inspection:
        raise HTTPException(status_code=404, detail="Inspection not found")
    
    # Get defects
    defects = db.query(DefectLog).filter(DefectLog.inspection_id == inspection_id).all()
    
    return {
        "inspection": inspection,
        "defects": defects,
        "defect_count": len(defects),
        "defect_rate": (float(inspection.quantity_rejected or 0) / float(inspection.quantity_inspected)) * 100 if inspection.quantity_inspected > 0 else 0
    }

@router.put("/inspections/{inspection_id}/approve")
async def approve_inspection(
    inspection_id: int,
    notes: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Approve inspection result"""
    inspection = db.query(QCInspection).filter(QCInspection.inspection_id == inspection_id).first()
    if not inspection:
        raise HTTPException(status_code=404, detail="Inspection not found")
    
    inspection.approved_by = current_user.user_id
    inspection.approved_at = datetime.utcnow()
    if notes:
        inspection.remarks = (inspection.remarks or "") + f"\nApproval notes: {notes}"
    
    db.commit()
    
    return {
        "message": "Inspection approved successfully",
        "inspection_number": inspection.inspection_number,
        "result": inspection.result
    }

# Defect Routes
@router.post("/defects")
async def log_defect(
    inspection_id: int,
    defect_type_id: int,
    quantity: float,
    severity: str,
    location: Optional[str] = None,
    description: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Log a defect"""
    defect = DefectLog(
        inspection_id=inspection_id,
        defect_type_id=defect_type_id,
        quantity=quantity,
        severity=severity,
        location=location,
        description=description
    )
    
    db.add(defect)
    db.commit()
    db.refresh(defect)
    
    return {
        "message": "Defect logged successfully",
        "defect_id": defect.defect_id
    }

@router.get("/defects")
async def list_defects(
    inspection_id: Optional[int] = None,
    defect_type_id: Optional[int] = None,
    severity: Optional[str] = None,
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """List defects with filters"""
    query = db.query(DefectLog)
    
    if inspection_id:
        query = query.filter(DefectLog.inspection_id == inspection_id)
    if defect_type_id:
        query = query.filter(DefectLog.defect_type_id == defect_type_id)
    if severity:
        query = query.filter(DefectLog.severity == severity)
    
    defects = query.order_by(DefectLog.logged_at.desc()).offset(skip).limit(limit).all()
    
    return {
        "total": query.count(),
        "defects": defects
    }

@router.get("/defects/types")
async def list_defect_types(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """List all defect types"""
    defect_types = db.query(DefectType).filter(DefectType.is_active == True).all()
    return {"total": len(defect_types), "defect_types": defect_types}

# Batch Approval Routes
@router.get("/batch-approval/{batch_number}")
async def get_batch_approval_status(
    batch_number: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get batch approval status"""
    approval = db.query(BatchApproval).filter(BatchApproval.batch_number == batch_number).first()
    
    if not approval:
        # Check if inspection exists
        inspection = db.query(QCInspection).filter(QCInspection.batch_number == batch_number).first()
        if not inspection:
            raise HTTPException(status_code=404, detail="Batch not found")
        
        return {
            "batch_number": batch_number,
            "status": "not_approved",
            "inspection": inspection
        }
    
    return {
        "batch_number": batch_number,
        "status": approval.status,
        "approval": approval
    }

@router.post("/batch-approval/{batch_number}")
async def approve_batch(
    batch_number: str,
    action: str,  # "approved", "rejected", "rework"
    notes: Optional[str] = None,
    rejection_reason: Optional[str] = None,
    rework_instructions: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Approve or reject batch"""
    # Get inspection for this batch
    inspection = db.query(QCInspection).filter(QCInspection.batch_number == batch_number).first()
    if not inspection:
        raise HTTPException(status_code=404, detail="No inspection found for this batch")
    
    # Check if already approved
    existing_approval = db.query(BatchApproval).filter(BatchApproval.batch_number == batch_number).first()
    
    if existing_approval:
        existing_approval.status = action
        existing_approval.approved_by = current_user.user_id
        existing_approval.approved_at = datetime.utcnow()
        existing_approval.notes = notes
        existing_approval.rejection_reason = rejection_reason
        existing_approval.rework_instructions = rework_instructions
    else:
        approval = BatchApproval(
            batch_number=batch_number,
            inspection_id=inspection.inspection_id,
            status=action,
            approved_by=current_user.user_id,
            approved_at=datetime.utcnow(),
            notes=notes,
            rejection_reason=rejection_reason,
            rework_instructions=rework_instructions
        )
        db.add(approval)
    
    db.commit()
    
    return {
        "message": f"Batch {action} successfully",
        "batch_number": batch_number,
        "status": action
    }
